### PR TITLE
Render refactor

### PR DESCRIPTION
**Pull request by jtlan**
Stage 1: renaming `_render()` to `render()`.

Next up: making `_doRender()` `protected` or `private`.

### CODE REVIEW

File: src/core/renderController.ts [hunk 1]
```diff
@@ -22,8 +22,8 @@ export module Core {
    * ```
    */
   export module RenderControllers {
-    var _componentsNeedingRender: {[key: string]: Component} = {};
-    var _componentsNeedingComputeLayout: {[key: string]: Component} = {};
+    var _componentsNeedingRender = new Utils.Set<Component>();
+    var _componentsNeedingComputeLayout = new Utils.Set<Component>();
```
Comment (aicioara) [inline src/core/renderController.ts]: Not part of this CR, but I thought we decided (or was brought up during the API refactor) that we instantiate all instance variables in the constructor as opposed to the declaration. Are we still iterating on that idea?

File: src/core/renderController.ts [hunk 2]
```diff
@@ -91,34 +91,23 @@ export module Core {
     export function flush() {
       if (_animationRequested) {
         // Layout
-        var toCompute = d3.values(_componentsNeedingComputeLayout);
-        toCompute.forEach((c) => c.computeLayout());
+        _componentsNeedingComputeLayout.values().forEach((c: Component) => c.computeLayout());
```
Comment (aicioara) [inline src/core/renderController.ts]: I really don't like single character variable names. We should probably move away from them

File: src/core/renderController.ts [hunk 3]
```diff
@@ -91,34 +91,23 @@ export module Core {
...
 
-        // Top level render.
-        // Containers will put their children in the toRender queue
-        var toRender = d3.values(_componentsNeedingRender);
-        toRender.forEach((c) => c._render());
+        // Top level render; Containers will put their children in the toRender queue
+        _componentsNeedingRender.values().forEach((c: Component) => c.render());
```
Comment (aicioara) [inline src/core/renderController.ts]: same

File: src/core/renderController.ts [hunk 4]
```diff
@@ -91,34 +91,23 @@ export module Core {
...
         _isCurrentlyFlushing = true;
-
-        // Finally, perform render of all components
-        var failed: {[key: string]: Component} = {};
-        Object.keys(_componentsNeedingRender).forEach((k) => {
+        var failed = new Utils.Set<Component>();
+        _componentsNeedingRender.values().forEach((c: Component) => {
           try {
-            _componentsNeedingRender[k]._doRender();
+            c._doRender();
```
Comment (aicioara) [inline src/core/renderController.ts]: same

File: src/core/renderController.ts [hunk 5]
```diff
@@ -91,34 +91,23 @@ export module Core {
...
           } catch (err) {
-            // using setTimeout instead of console.log, we get the familiar red
-            // stack trace
-            setTimeout(() => {
-              throw err;
-            }, 0);
-            failed[k] = _componentsNeedingRender[k];
+            // throw error with timeout to avoid interrupting further renders
+            window.setTimeout(() => { throw err; }, 0);
```
Comment (aicioara) [inline src/core/renderController.ts]: no need for `{` `}`

Comment (aicioara) [inline src/core/renderController.ts]: Does this pass tslint? I thought we enforce newline after `{`. I prefer the old way.

**Comment by aicioara** [comment]
Minor style things. Mainly personal preference.

File: src/core/renderController.ts (hunk 1 above)
Comment (jtlan) [inline src/core/renderController.ts]: This isn't a `class`, so there's no `constructor`. The `module` just happens to have some variables in it.

File: src/core/renderController.ts (hunk 2 above)
Comment (jtlan) [inline src/core/renderController.ts]: Okay, I'll make the change.

File: src/core/renderController.ts (hunk 1 above)
Comment (aicioara) [inline src/core/renderController.ts]: Fair enough

File: src/core/renderController.ts (hunk 2 above)
Comment (bluong) [inline src/core/renderController.ts]: If you're worried about character constraints, feel free to remove the typing at the function declaration stage.  We should probably do this moving forward

**Comment by aicioara** [comment]
Also, to save you some time with the build, you should also cherry-pick my first commit from here https://github.com/palantir/plottable/pull/1967

**Comment by aicioara** [comment]
@jtlan you need another merge

**Comment by aicioara** [comment]
Once the build passes

**Pull request merged**